### PR TITLE
Add CI for Linux ARM64

### PR DESCRIPTION
**Pull request by martin-g**
Remove '-m64' CXX flag. It is implicit for x86_64. 
Change '-march=' to 'native', so that it will build with all available CPU features on the build system. This way Paw could be build for aarch64 too

### CODE REVIEW

**Comment by martin-g** [comment]
This PR is similar to https://github.com/hannespetur/paw/pull/15 but against `dev` branch.

**Comment by martin-g** [comment]
@hannespetur Any feedback on these changes ?
Also could you please enable the CI workflow for this PR so that we see it passes ? 
Thank you!

**Comment by hannespetur** [comment]
Hi @martin-g , thank you for working on this nice addition! I am on a paternal leave, so apologies for the lack of response from my side. 

One thing that should be considered is that paw::Align (and therefore graphtyper) compile with many different set of CPU extensions, like SSE4.1, AVX, AVX2 and so on. At runtime the "best" extension set is chosen. I don't know much about ARM64 architecture, which CPU extensions are normally found on those systems? Because those sets should be compiled as well.

Edit: And actually this is a reason I have been avoiding "-march=native". Because we often compile on a system but run on another. Would it be ok if we made CMake detect if it's being compiled on ARM64 and then only use "-march=native" in that case? 

Best,
Hannes

**Comment by martin-g** [comment]
Hi @hannespetur !
**I** should apology for disturbing your parental leave!

TL;DR
I will rework the PR to use the same approach as in Graphtyper:
https://github.com/DecodeGenetics/graphtyper/pull/114/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR145-R155

This way the user can decide what arch related compilation options to pass.

----------------------------------------------------

> which CPU extensions are normally found on those systems?

The related extension is NEON:

https://github.com/p12tic/libsimdpp/blob/9dac213d6965bf57c7accf9dda6d16c3bed0e3ac/simdpp/dispatch/get_arch_linux_cpuinfo.h#L36-L42

I will add an entry for it at https://github.com/hannespetur/paw/blob/master/src/align.cpp !

----------------------------------------------------

Using `-march=native` is not good for distribution! But using a generic value (like `-march=x86_64`) might be sub-optimal if you build on a machine with less extensions than the one where executing. But making it possible to specify the options via CMAKE option should be flexible enough for all use cases.

**Comment by martin-g** [comment]
All improvements are done!

There is something that still bothers me though - `cmake`s output contains this on both my ARM64 VM and at CI (QEMU emulated):
```
...
-- Performing Test CAN_RUN_ARM_NEON
-- Performing Test CAN_RUN_ARM_NEON - Failed
-- Performing Test CAN_RUN_ARM64_NEON
-- Performing Test CAN_RUN_ARM64_NEON - Failed
...
-- Compilable archs are: NONE_NULL;ARM64_NEON
-- Runnable archs are: NONE_NULL
...
```

I don't understand why `ARM64_NEON` is not in the `RUNNABLE_ARCHS` list.

**Comment by hannespetur** [comment]
Awesome.

> Using `-march=native` is not good for distribution! But using a generic value (like `-march=x86_64`) might be sub-optimal if you build on a machine with less extensions than the one where executing. But making it possible to specify the options via CMAKE option should be flexible enough for all use cases.

Yes, I agree. It should be set as an option with `CMAKE_CXX_FLAGS`.  

> I don't understand why ARM64_NEON is not in the RUNNABLE_ARCHS list.

Hmm, weird. The code it is trying to run is here: https://github.com/hannespetur/paw/blob/dev/cmake/SimdppMultiarch.cmake#L746

Does at least get_current_arch() find ARM64_NEON? You can see from the output of paw align example 2.

Best,
Hannes

**Comment by hannespetur** [comment]
I think the CI should now run for you without my approval.

**Comment by martin-g** [comment]
align example 2 fails on ARM64 with:
```
...
[ 20%] Linking CXX executable ../../bin/align_ex2
/usr/bin/ld: CMakeFiles/align_ex2.dir/align_ex2_libsimdpp.cpp.o: in function `main':
/home/mgrigorov/git/hpc/paw/examples/align/align_ex2_libsimdpp.cpp:75: undefined reference to `void paw::arch_sse2::pairwise_alignment<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned short>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, paw::AlignmentOptions<unsigned short>&)'
collect2: error: ld returned 1 exit status
make[2]: *** [examples/align/CMakeFiles/align_ex2.dir/build.make:99: bin/align_ex2] Error 1
make[1]: *** [CMakeFiles/Makefile2:304: examples/align/CMakeFiles/align_ex2.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
```

It hardcodes `sse2` ...
I will see how to rework it!

**Comment by martin-g** [comment]
The problem with the `RUNNABLE_ARCHS` seems to be [here](https://github.com/hannespetur/paw/blob/82f12b9a496218b480cc77f72b17bba559b9ca00/cmake/SimdppMultiarch.cmake#L762-L779)
It compiles fine but it seems there is a problem while running it ...

I've compared it against https://github.com/p12tic/libsimdpp/blob/master/cmake/SimdppMultiarch.cmake and the ARM64 related code is the same.

**Comment by martin-g** [comment]
I've extracted https://github.com/hannespetur/paw/blob/82f12b9a496218b480cc77f72b17bba559b9ca00/cmake/SimdppMultiarch.cmake#L747-L779 to a file and:

```
$ g++ ../arm64.c && ./a.out
../arm64.c: In function ‘int main()’:
../arm64.c:22:36: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
   22 |         uint32x4_t one = vld1q_u32((uint32_t*)(*p));
      |                                    ^~~~~~~~~~~~~~~
../arm64.c:26:19: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
   26 |         vst1q_u32((uint32_t*)(*p), one);
      |                   ^~~~~~~~~~~~~~~
fish: Job 1, './a.out' terminated by signal SIGSEGV (Address boundary error)

```

But if I replace it with https://github.com/hannespetur/paw/blob/82f12b9a496218b480cc77f72b17bba559b9ca00/cmake/SimdppMultiarch.cmake#L693-L728 then all is fine and the output is:
```
-- Compilable archs are: NONE_NULL;ARM64_NEON
-- Runnable archs are: NONE_NULL;ARM64_NEON
```

Unfortunatelly https://github.com/p12tic/libsimdpp looks unmaintained.

**Comment by martin-g** [comment]
Filled https://github.com/p12tic/libsimdpp/issues/159

**Comment by martin-g** [comment]
@hannespetur I've debugged the issue with libsimdpp - https://github.com/martin-g/libsimdpp/commit/b2be50d8f94fb65788af874512e7a72babbcf502

PR for upstream: https://github.com/p12tic/libsimdpp/pull/160 (**merged**!)

Now the output is:
```
...
-- Compilable archs are: NONE_NULL;ARM64_NEON
-- Runnable archs are: NONE_NULL;ARM64_NEON
...
```

**Comment by markjens** [comment]
Awesome work, @martin-g !
I hope your improvements will be merged soon!
Until then we will use a build from your fork.

**Review by hannespetur — APPROVED** [review]

**Comment by hannespetur** [comment]
Thank you @martin-g for your work and patience.
